tensorflow-keras/src/main/java/org/tensorflow/keras/backend/tf/NN.java

```diff
@@ -0,0 +1,134 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+=======================================================================*/
+package org.tensorflow.keras.backend.tf;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import org.tensorflow.Operand;
+import org.tensorflow.ndarray.Shape;
+import org.tensorflow.op.Op;
+import org.tensorflow.op.Ops;
+import org.tensorflow.op.nn.SparseSoftmaxCrossEntropyWithLogits;
+import org.tensorflow.types.TBfloat16;
+import org.tensorflow.types.TFloat16;
+import org.tensorflow.types.TFloat32;
+
+/** NN Operations */
+public class NN {
+
+  /**
+   * Computes sparse softmax cross entropy between `logits` and `labels`.
+   *
+   * @param tf
+   * @param labels `Tensor` of shape `[d_0, d_1, ..., d_{r-1}]` (where `r` is rank of `labels` and
+   *     result) and dtype `int32` or `int64`. Each entry in `labels` must be an index in `[0,
+   *     num_classes)`. Other values will raise an exception when this op is run on CPU, and return
+   *     `NaN` for corresponding loss and gradient rows on GPU.
+   * @param logits Per-label activations (typically a linear output) of shape `[d_0, d_1, ...,
+   *     d_{r-1}, num_classes]` and dtype `float16`, `float32`, or `float64`. These activation
+   *     energies are interpreted as unnormalized log probabilities.
+   * @return A `Tensor` of the same shape as `labels` and of the same type as `logits` with the
+   *     softmax cross entropy loss.
+   */
+  public static Operand sparse_softmax_cross_entropy_with_logits(
+      Ops tf, Operand labels, Operand logits) {
+    // assert shapeIsCompatible(labels.asOutput().shape(), logits.asOutput().shape()):
+    //        String.format("Shapes %s and %s are incompatible",
+    //                labels.asOutput().shape(), logits.asOutput().shape());
+    tf = tf.withSubScope("SparseSoftmaxCrossEntropyWithLogits");
+    Operand precise_logits = logits;
+    boolean convertToFloat32 =
+        logits.asOutput().dataType() == TFloat16.DTYPE
+            || logits.asOutput().dataType() == TBfloat16.DTYPE;
```

Comment:
This might get out of date if we add more float types in the future. We should add a type family called `TFloating` that is used to tagged all our floating point tensor types (that is what I did in https://github.com/karllessard/tensorflow-java/tree/tensor-as-ndarrays-3)

Reply:
This specific logic is primarily checking to see if the float type is smaller than `TFloat32` and if so, flag it for casting up later. I don't thing `TFloating` would capture this specific logic.  I have modified my version of DataType to add `isFloating()`, `isInteger()`, `isNumeric()`,` isBoolean()`, and` isString()`. I dif this based on Craig's suggestion earlier.

also sparseCrossEntropyWithLogits is defined in nn in Python TF "nn.softmax_cross_entropy_with_logits". It has nothing to do with SparseTensor if that is what you thought.